src/Servers/Kestrel/Core/src/Internal/Infrastructure/StringUtilities.cs

```diff
@@ -118,39 +118,35 @@ public static unsafe bool TryGetAsciiString(byte* input, char* output, int count
         /// <param name="separator"></param>
         /// <param name="number"></param>
         /// <returns></returns>
-        public static unsafe string ConcatAsHexSuffix(string str, char separator, uint number)
+        public static string ConcatAsHexSuffix(string str, char separator, uint number)
         {
             var length = 1 + 8;
             if (str != null)
             {
                 length += str.Length;
             }
 
-            // stackalloc to allocate array on stack rather than heap
-            char* charBuffer = stackalloc char[length];
-
-            var i = 0;
-            if (str != null)
+            return string.Create(length, (str, separator, number), (charBuffer, tuple) =>
             {
-                for (i = 0; i < str.Length; i++)
+                var i = 0;
+                if (tuple.str != null)
                 {
-                    charBuffer[i] = str[i];
+                    for (i = 0; i < tuple.str.Length; i++)
+                    {
+                        charBuffer[i] = tuple.str[i];
+                    }
                 }
-            }
-
-            charBuffer[i] = separator;
-
-            charBuffer[i + 1] = _encode16Chars[(int)(number >> 28) & 0xF];
-            charBuffer[i + 2] = _encode16Chars[(int)(number >> 24) & 0xF];
-            charBuffer[i + 3] = _encode16Chars[(int)(number >> 20) & 0xF];
-            charBuffer[i + 4] = _encode16Chars[(int)(number >> 16) & 0xF];
-            charBuffer[i + 5] = _encode16Chars[(int)(number >> 12) & 0xF];
-            charBuffer[i + 6] = _encode16Chars[(int)(number >> 8) & 0xF];
-            charBuffer[i + 7] = _encode16Chars[(int)(number >> 4) & 0xF];
-            charBuffer[i + 8] = _encode16Chars[(int)number & 0xF];
-
-            // string ctor overload that takes char*
-            return new string(charBuffer, 0, length);
+                
+                charBuffer[i + 8] = _encode16Chars[(int)tuple.number & 0xF];
```

Comment:
I'm not sure (and haven't checked) whether the JIT will apply CSE on `tuple.number`. Therefore I would read this once into a local, then use the local. So
```c#
char[] encode16Chars = _encode16Chars;
uint value = tuple.number;
charBuffer[i + 8] = encode16Chars[(value >> 4) & 0xF];
// ...
```

Reply:
Or it is better (and more readable?) to deconstruct the tuple, to get the locals from the compiler.